### PR TITLE
Re-add a unify mode using -P

### PR DESCRIPTION
**Pull request by Meinersbur**
This is a proof-of-concept of using the `-P` preprocessor flag in cpp2 mode (i.e. only use preprocessor output for has lookup) as outlined in https://github.com/ccache/ccache/issues/812#issuecomment-786519591. Since ccache uses line directives inserted by the preprocessor to determine which files are to be hashed in direct mode, we have to get this information another way. It is done by instructing the preprocessor to emit a `.d` file and which is parsed the same way as depend mode does.

The motivation of this patch is to increase the chances of a cache hit after only whitespace changes. In particular, removing the line directives may ignore added/removed empty or comment comment lines from the source input. It was possible to ignore even more whitespace changes with `CCACHE_UNIFY` option, that unfortunately was removed in ccache 3.7.7 (https://github.com/ccache/ccache/commit/0774bfe4c71b342822170057df0aef15d371d892).

What kinds of whitespace changes can be ignored depends on the compiler: gcc's `-E -P` mode removes all empty lines, but keeps indention and only normalizes horizontal whitespace to at most a single space. clang's `-E -P` mode even only discards empty lines if there are more than 8 of them. However, I implemented a `-fnormalize-whitespace` mode in clang (https://github.com/meinersbur/llvm-project/tree/normalize-whitespace) that basically does what ccache's unify did, ignoring almost all formatting changes. If it finds use I would upstream this to llvm project.

This approach holds several advantages:
 * Preprocessed file does not need to be parsed, just to be hashed.
 * All compilers with a gcc-compatible command line interface have to emit a make-compatible dependency file, while the line-directive is compiler-specific.
 * Smaller `.i`/`.ii` temporary file, less to hash
 * Ignores more irrelevant source changes, such as adding/removing empty lines or comments, indention
 * Can handle `-P` being passed as an argument (#812)

Most problems of `CCACHE_UNIFY` do not apply, only the following disadvantages remain:
 * Compiler diagnostics may be off if e.g. only an empty line is added or removed from the source.
 * Also, debug info line can be off.
 * `__LINE__` macros (e.g. in an `assert`) will cause a miss if the number of lines before changed.

The last is probably what you want, but may reduce the usefulness of this mode. One might want to use an `assert` that does not print the line. The first two can be mitigated by not using this mode when `-g` is pass and not caching if there is any diagnostic output. Direct mode hits are still possible. A sloppiness option could still enable `-P`-mode nonetheless. Even though `-fnormalize-whitespace` is only supported with clang, the mode could still by uses by pipeing through `clang -fpreprocessed -fnormalize-whitespace` or by linking libclang into ccache.

This patch is a draft, I only tested it with my workflow. Missing changes in the documentation, statistics, configuration and different compilers and modes (such as invoking only the preprocessor without compilation) may not work. If the project is interested in merging such a patch, I'd appreciate hints on what still needs to be done or if the maintainers take over from here. It helps me being less hesitant to clang-format or edit comments in central header files.

### CODE REVIEW

**Comment by jrosdahl** [comment]
Thanks for the idea.

I would suggest changing the title to something like "Re-add a unify mode using -P" so that it's easier to understand the intent.

I'm OK with the idea in principle if the implementation can be made correct without too much added complexity. The devil's in the details as can be witnessed by many tests failing with the current POC.

> Preprocessed file does not need to be parsed, just to be hashed.

That's not quite true. The preprocessed file should still be scanned for `.incbin` and log messages from the distcc-pump wrapper.

> All compilers with a gcc-compatible command line interface have to emit a make-compatible dependency file, while the line-directive is compiler-specific.

Right. Note though that ccache supports compilers without 100% GCC-like behavior for dependency files, most notably EDG-based compilers (such as the Green Hills compiler, GHS). I think that the mode has to be opt in.

> Can handle -P being passed as an argument (#812)

As noted in https://github.com/ccache/ccache/issues/812#issuecomment-797035331, I currently don't think it's terribly important to support `-P`.

**Comment by Meinersbur** [comment]
> I would suggest changing the title to something like "Re-add a unify mode using -P" so that it's easier to understand the intent.

I changed the title to you suggestion but I think it is misleading because without `-fnormalize-whitespace` the unification is incomplete.


> > Preprocessed file does not need to be parsed, just to be hashed.
> 
> That's not quite true. The preprocessed file should still be scanned for `.incbin` and log messages from the distcc-pump wrapper.

Incbin is non-standard and I don't think a lot of projects are using it. The more compatible options would be something like `xxd -i`, `bin2c` or `ld -b binary`. C++23 is considering adding [`std::embed`](http://open-std.org/JTC1/SC22/WG21/docs/papers/2018/p1040r0.html) or `#embed` to the standard, and they want to ensure that `-MD`/`-MMD` will be able to include the filename.

However, for safety and distcc-pump I agree we may still need to parse the file, although I would add a sloppiness option to disable that scanning for projects that know they don't use incbin. I could try implementing detection of `.incbin` dependency in clang. Note that the current detection is not 100% safe since one can hide the `.incbin` command like `__asm__(".inc" "bin")`, depends-mode will miss it, and while the blake3 implementation is efficient and vectorized, the parsing code is not.


> Right. Note though that ccache supports compilers without 100% GCC-like behavior for dependency files, most notably EDG-based compilers (such as the Green Hills compiler, GHS). I think that the mode has to be opt in.

OK, I don't know these compilers. What if gcc/clang was detected?

**Comment by Meinersbur** [comment]
What should the option enabling this be called? `unify_mode`?

**Comment by jrosdahl** [comment]
> I changed the title to you suggestion but I think it is misleading because without `-fnormalize-whitespace` the unification is incomplete.

Even if it's not functionally equivalent to the old unify mode, I guess it still has to be called _something_ (perhaps not in the PR but at least in the documentation) and I thought that "a unify mode" would be fluffy enough to hint that it's not "_the_ unify mode". Better name suggestions are of course welcome.

> Incbin is non-standard and I don't think a lot of projects are using it.

I think that the Linux kernel project motivates the detection code just fine even though it's only one project (see #136).

> The more compatible options would be something like `xxd -i`, `bin2c` or `ld -b binary`.

If you convince all ccache users that they should use something else then we can for sure drop the detection. :wink:

> [...] However, for safety and distcc-pump I agree we may still need to parse the file, although I would add a sloppiness option to disable that scanning for projects that know they don't use incbin.

That would be perfectly fine.

> I could try implementing detection of `.incbin` dependency in clang.

Not sure I follow here. If you suggest depending on a potential future implementation in Clang I think I would need more sales pitches.

> Note that the current detection is not 100% safe since one can hide the `.incbin` command like `__asm__(".inc" "bin")`

Right, and then the user deserves to get a false cache hit.

> depends-mode will miss it

That's a good point. I would consider that a bug in the depend mode. For the record, I'm not convinced that I made the right choice to accept the direct mode into ccache since it adds complexity and is a bit hard to reason about, but as long as it's opt-in it doesn't do too much harm.

> and while the blake3 implementation is efficient and vectorized, the parsing code is not.

Of course.

> OK, I don't know these compilers. What if gcc/clang was detected?

Doesn't it need to be opt in anyway because of the issues with line numbers in compiler diagnostics, assert macros, etc being potentially off?

> What should the option enabling this be called? unify_mode?

I don't have any better suggestion at the moment.

**Comment by Meinersbur** [comment]
The patch is ready for review. If preferred, I could extract parts of it into separate pull requests.

This assumes that support for the `-fnormalize-whitespace` will be accepted before the Clang 13 release (ttps://reviews.llvm.org/D104601). I suggest to wait until that before merging.

**Comment by jrosdahl** [comment]
Thanks.

I just wanted to say that I haven't forgotten about this PR, but I'm currently focusing on [storage backends](https://github.com/ccache/ccache/projects/1) so I'll have to come back to this at some later time.

**Comment by Meinersbur** [comment]
OK, usefulness is limited before we get `-fnormalize-whitespace` anyway.

**Comment by Meinersbur** [comment]
The feature was accepted in clang from clang 14.0.0 (NOT the next release) as `-fminimize-whitespace`. 

Note I rebase only to the commit before accd21e452da13f12877571f90dae7e34feca23b. That commit makes unittests read the user's configuration in `.ccache/ccache.conf` and cause it fail if it diverges from the default values.

**Comment by jrosdahl** [comment]
> Note I rebase only to the commit before [accd21e](https://github.com/ccache/ccache/commit/accd21e452da13f12877571f90dae7e34feca23b). That commit makes unittests read the user's configuration in `.ccache/ccache.conf` and cause it fail if it diverges from the default values.

Thanks for noting this. Fixed in a2e6316eae064cb283beb481d2135f5188d0bd9a.

**Comment by Meinersbur** [comment]
I rebased everything to the current master branch. I would appreciate a review.

**Comment by Meinersbur** [comment]
ping

**Comment by Meinersbur** [comment]
ping

**Comment by jrosdahl** [comment]
The status is still that I haven't forgotten about this PR. No ping noise needed or appreciated. Since this PR has become more complex than I initially thought it would I haven't had time yet to think things through enough to give good feedback.

File: doc/MANUAL.adoc [hunk 1]
```diff
@@ -494,6 +494,15 @@ _<<Using ccache with other compiler wrappers>>_.
...
     Clang-based compiler.
+[#config_compiler_type_clang_minimize_whitespace] 
+*clang_normalize_whitespace*::
+    Clang starting at version 14 supports the *-fminimize-whitespace*
+    option. The switch removes as much whitespace between tokens as
+    possible which reduces the chance of hashes of preprocessed
+    output being different only because of formatting changes. It is most
+    useful in <<the_unify_mode,unify mode>> where the *-P* switch
+    also removes newlines as much as possible, otherwise it only tries
+    to remove spaces between tokens on the same line.
```
[the remark this answers is not in the document]
Comment (Meinersbur) [reply doc/MANUAL.adoc:505]: `compiler_type` seemed to be ccache's mechanism to identify a the set of flags supported by a compiler, `-fminimize-whitespace` would be one flavor of clang and therefore the most straightforward way to add another supported command line set. If gcc/others start supporting the flag, it could be made a bitflag, but at this point we cannot assume that gcc would also call there command line flag `-fminimize-whitespace`, but maybe `-fminify` in which case we'd indeed need a `gcc_minify`. What gcc will do is speculation at this point. There could be a more general mechanism to handle different versions of a compiler, but that's a lot more than needed for `-fminimize-whitespace` to work.

sccache compiles a small program to determine how the compiler identifies itself, including its version. The ccache way to do this seemed to be to follow the symlink. Clang's `bin/lang` and `bin/clang++` are symlinks to `bin/clang-${version}`, i.e. sufficient for our purpose. Unfortunately I only recently discovered that at least Ubuntu changed this in their distribution and does this the other way around: `clang-14` is a symlink to `/usr/lib/llvm-14/bin/clang`, so this patch at least needs to be extended to detect that pattern as well.

The best solution that also works with any compiler derived from clang would be a test invocation of `$cc -P -fminimize-whitespace` to see whether specifically this flag is support, but I assumed we'd not want to pay the cost of an additional program execution every time and I am not knowledgeable enough to add caching for the result.

When using unify mode there is no reason to no use `-fminimize-whitespace`, it is what the original unify mode did. Adding a configuration for enable it would make the ccache configuration more complex and most users would miss out on its advantages if had to be enabled explicitly.

File: doc/MANUAL.adoc [hunk 2]
```diff
@@ -800,6 +809,50 @@ set <<config_temporary_dir,*temporary_dir*>> since ccache will fail to create
 temporary files otherwise. You may also want to set <<config_stats,*stats*>> to
 *false* make ccache not even try to update stats files.
 
+[#config_preprocessor]
+*preprocessor* (*CCACHE_PREPROCESSOR*)::
```
[the remark this answers is not in the document]
Comment (Meinersbur) [reply doc/MANUAL.adoc:813]: I am using it when using gcc as a compiler, so I can use clang to identify whether there is a change (i.e. using `-fminimize-whitespace`).

But I understand that the situation might not be beneficial enough to be worth the complexity. It least I can move it to another PR.

File: doc/MANUAL.adoc [hunk 3]
```diff
@@ -949,6 +1058,11 @@ filesystem as the `CCACHE_DIR` path, but this requirement has been relaxed.
     in the cache directory. This is mostly useful when you wish to share your
     cache with other users.
 
+[#config_unify_mode]
+*unify_mode* (*CCACHE_UNIFY* or *CCACHE_NOUNIFY*, see _<<Boolean values>>_ above)::
+  
+    If true, the unify mode will be used. The default is true. See
```
[the remark this answers is not in the document]
Comment (Meinersbur) [reply doc/MANUAL.adoc:1064]: Correct.

File: doc/MANUAL.adoc [hunk 4]
```diff
@@ -1356,10 +1480,12 @@ algorithm, to detect corruption.
...
 cache:
 
-* the *preprocessor mode*, where ccache runs the preprocessor on the source
-  code and hashes the result
 * the *direct mode*, where ccache hashes the source code and include files
   directly
+* the *preprocessor mode*, where ccache runs the preprocessor on the source
+  code and hashes the result
+* the *unify mode*, where ccache additionally does not try to preserve
+  line breaks such that formatting changes do not trigger recompilation
```
[the remark this answers is not in the document]
Comment (Meinersbur) [reply doc/MANUAL.adoc:1488]: It's a "mode" so seems to belong to this list. However, I think "mode" is not a good terminology, which seems to imply that these are mutually exclusive, but you still get "direct mode" hits in either "preprocessor mode" or "direct mode", nor "direct mode" only.

File: doc/MANUAL.adoc [hunk 5]
```diff
@@ -1475,6 +1601,80 @@ The depend mode will be disabled if any of the following holds:
...
+
+Unify mode is most powerful when used in conjunction with Clang's
+<<config_compiler_type_clang_minimize_whitespace,*-fminimize-whitespace*>>.
+This flag normalizes all line line breaks unless semantically required
+(e.g. after a #pragma or inside a raw string). Otherwise, with *-P*
+only, the preprocesser still keeps those tokens on the same line how
+they are in the input file. The handling of empty lines of lines with
+only comments depends on the compiler. Note that in any preprocessor
+mode (including unify mode), column information (*-gcolumn-info*,
+enabled in gcc by default) is never guaranteed to match.
```
[the remark this answers is not in the document]
Comment (Meinersbur) [reply doc/MANUAL.adoc:1624]: Preprocessor output (-E) cannot preserve columns for each token. For instance the `TYPE` in
```
TYPE varname;
```
may be expanded to `double` such that `varname` will not be in column 6 after expansion. However, clang (and gcc) try to preserve the column of the first token of each line (i.e. indention), but will preserve only a single space between other tokens on the same line.

Therefore you can get hash match after preprocessing with the following:
```
TYPE                     varname;
```
which both are preprocessed to:
```
double varname;
```

Note that gcc and clang to emit diagnostics showing from which macro a token is expanded from, therefore that diagnostic will be different in `no_second_cpp`. mode.

File: doc/MANUAL.adoc [hunk 6]
```diff
@@ -906,13 +987,41 @@ Examples:
...
+    debuggers show the wrong line number or cannot set breakpoint at
+    specific line numbers.
++
+If <<config_run_second_cpp,run_second_cpp>> is disabled, the debug
+info will instead point to the file containing the preprocessed source.
+The file is temporary and will have been deleted already which makes the
+debug info rather useless.
+
+[#config_sloppy_unify_with_output]
+*unify_with_output*::
```
[the remark this answers is not in the document]
Comment (Meinersbur) [reply doc/MANUAL.adoc:1015]: Agreed.

File: doc/MANUAL.adoc [hunk 7]
```diff
@@ -881,6 +934,34 @@ Examples:
...
+    <<the_unify_mode,unify mode>>, <<the_direct_mode,*direct_mode*>> and
+    nearly every buildsystem).
++
+The incbin sloppiness option allows ccache to skip scanning for
+the .incbin directive. This is mostly beneficial in
+<<the_unify_mode,unify mode>> where searching for *.incbin* is the only
+reason for it to parse the preprocessed file and therefore could be skipped.
++
+Neither <<_the_direct_mode,direct mode>> nor
+<<the_depend_mode,depend mode>> take *.incbin*
```
[the remark this answers is not in the document]
Comment (Meinersbur) [reply doc/MANUAL.adoc:960]: If the file gets into the cache via depend mode, the manifest will not include the `.incbin` file and therefore get a direct hit in direct mode even though the `.incbin` file may have changed. Therefore I think wording is correct, direct mode does not look for `.incbin` instruction, it just hashes the files in the manifest.

This might be an issue with calling them "modes".

File: doc/MANUAL.adoc [hunk 8]
```diff
@@ -1475,6 +1601,80 @@ The depend mode will be disabled if any of the following holds:
...
+documented, combining it with unify mode is not recommended. Together
+with removing line information from the preprocessed output, it also
+does not preserve which declarations are from a system header. Compiler
+behaviour may differ in system headers, in particular compiler
+diagnostics are suppressed when originating from a system header.
+That is, with <<config_sloppy_unify_with_output,*unify_with_output*>>
+sloppiness option, this results in noisy compiler diagnostics.
+Without <<config_sloppy_unify_with_output,*unify_with_output*>>, any
+compiler output will trigger a fallback compiler invocation using the
+original files and without caching.
```
[the remark this answers is not in the document]
Comment (Meinersbur) [reply doc/MANUAL.adoc:1676]: There didn't seem a good reason to artificially block this combination of configuration options if the user knows what they are doing. However, I also don't have a strong desire to allow it.

What should happen with contradicting configuration options? Error? Warning? Silently ignore?

File: src/third_party/nonstd/optional.hpp [hunk 9]
```diff
@@ -1198,6 +1198,7 @@ class optional
     )
     operator=( optional && other ) noexcept
     {
+	reset();
```
[the remark this answers is not in the document]
Comment (Meinersbur) [reply src/third_party/nonstd/optional.hpp:1201]: At some point I found that the previous state of an optional was leaking after assignment. I don't remember the details, I will remove it until I rediscover the problem.

File: src/ccache.cpp [hunk 10]
```diff
@@ -1036,8 +1161,47 @@ get_result_key_from_cpp(Context& ctx, Args& args, Hash& hash)
...
     size_t args_added = 2;
     args.push_back("-E");
+
+    // Use -fminimize-whitespace if available. Outside unify mode, this will
+    // still normalize tokens on this same line.
+    if (ctx.config.compiler_type() == CompilerType::clang_minimize_whitespace
+        && language_supports_unify(ctx.args_info.actual_language)) {
+      args.push_back("-fminimize-whitespace");
+      args_added += 1;
+    }
```
[the remark this answers is not in the document]
Comment (Meinersbur) [reply src/ccache.cpp:1176]: Column information is already inaccurate even without `-fminimize-whitespace` as detailed in the docs. What changes with `-fminimize-whitespace`:
 * Clang tries to preserve indentation of the first token. That is, your example would indeed show a different column, but only because it is the first token on line 3.
 * Clang emits a single space between two tokens on the same line if there was a space before the second token. With `-fminimize-whitespace`, a space is only added if required to not merge two tokens into one.

File: src/argprocessing.cpp [hunk 11]
```diff
@@ -1206,6 +1207,21 @@ process_args(Context& ctx)
...
+    ctx.config.set_unify_mode(false);
+  }
+  if (ctx.config.unify_mode() && ctx.args_info.seen_MMD
+      && !ctx.config.sloppiness().is_enabled(core::Sloppy::system_headers)) {
+    LOG_RAW(
+      "Disabling unify mode: -MMD output requested, but -MD output needed to "
+      "track system headers. Either enable the system_header sloppiness option "
+      "or use run_second_cpp");
+    ctx.config.set_unify_mode(false);
+  }
```
[the remark this answers is not in the document]
Comment (Meinersbur) [reply src/argprocessing.cpp:1223]: `-H` intermixes diagnostics with included files, or we need another compiler invocation. The logic `-MD`/`-MMD` for handling depfiles already exists, adding support for `-H` would be new feature unrelated to unify mode.

File: src/third_party/nonstd/optional.hpp (hunk 9 above)
[the remark this answers is not in the document]
Comment (Meinersbur) [reply src/third_party/nonstd/optional.hpp]: The problem is the following waring by gcc:
```
/home/runner/work/ccache/ccache/src/third_party/nonstd/optional.hpp: In function ‘nonstd::expected_lite::expected<Digest, {anonymous}::Failure> to_cache(Context&, Args&, nonstd::optional_lite::optional<Digest>, const Args&, Hash*)’:
/home/runner/work/ccache/ccache/src/third_party/nonstd/optional.hpp:1203:78: error: ‘*((void*)& __tmp +28)’ may be used uninitialized in this function [-Werror=maybe-uninitialized]
         else if ( (has_value() == true ) && (other.has_value() == true ) ) { contained.value() = std::move( *other ); }
```

**Comment by jrosdahl** [comment]
The status is still that I haven't forgotten about this PR, but I consider it low priority (sorry) and I don't have time to engage with the topic.

**Comment by jrosdahl** [comment]
Closing this now since it won't be merged in its current form.
